src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -175,6 +195,97 @@ public Object invokeMethod(String name, Object args) {
+            // in this case we assume this is building a nested object used as an eventual
+            // parameter of an executable symbol, e.g.,
+            //
+            // hg source: 'https://whatever/', clean: true, browser: kallithea('https://whatever/')
+
+            // also note that in this case 'd' is not trustworthy, as depending on
+            // where this UninstantiatedDescribable is ultimately used, the symbol
+            // might be resolved with a specific type.
+            return ud;
+        } else {
+            try {
+                // execute this Describable through a meta-step
+
+                // split args between MetaStep (represented by mm) and Describable (represented by dm)
+                DescribableModel<?> mm = new DescribableModel(metaStep.clazz);
+                DescribableModel<?> dm = new DescribableModel(d.clazz);
+                DescribableParameter p = mm.getFirstRequiredParameter();
+
+                // order of preference:
+                //      1. mandatory parameter in mm
+                //      2. mandatory parameter in dm
+                //      3. other parameters in mm
+                //      4. other parameters in dm
+                // mm is preferred over dm because that way at least the arguments that dm defines
+                // act consistently
+                Map<String,Object> margs = new TreeMap<>();
+                Map<String,Object> dargs = new TreeMap<>();
+                for (Entry<String, ?> e : ud.getArguments().entrySet()) {
+                    String n = e.getKey();
+                    Object v = e.getValue();
+                    DescribableParameter mp = mm.getParameter(n);
+                    DescribableParameter dp = dm.getParameter(n);
+
+                    if (mp!=null && mp.isRequired()) {
+                        margs.put(n,v);
+                    } else
+                    if (dp!=null && dp.isRequired()) {
+                        dargs.put(n,v);
+                    } else
+                    if (mp!=null) {
+                        margs.put(n,v);
+                    } else {
+                        // dp might be null, but this error will be caught by UD.instantiate() later
+                        dargs.put(n,v);
+                    }
+                }
+
+                ud = new UninstantiatedDescribable(symbol, null, dargs);
+                margs.put(p.getName(),ud);
+
+                return invokeStep(metaStep,new NamedArgsAndClosure(margs,args.body));
+            } catch (Exception e) {
+                throw new IllegalArgumentException("Failed to prepare "+symbol+" step",e);
+            }
+        }
+    }
+
+    /**
+     * Finds a meta step that can handle a Describable of the given type 'd'
+     */
+    private StepDescriptor findMetaStep(Descriptor d) {
+        OUTER:
+        for (StepDescriptor sd : StepDescriptor.allMeta()) {
+            if (sd.getMetaStepArgumentType().isAssignableFrom(d.clazz)) {
```

Comment:
Seems like it would be simpler to delete `allMeta` and do the null check here instead.